pkg/ruler/manager_metrics.go

```diff
@@ -271,3 +273,42 @@ func (m *RuleEvalMetrics) deletePerUserMetrics(userID string) {
 		m.RulerQuerySeconds.DeleteLabelValues(userID)
 	}
 }
+
+type RuleGroupMetrics struct {
+	mtx               sync.Mutex
```

Comment:
We need to because we have unit tests for `UpdateRuleGroupsInStore`. Without this mutex the test will fail when run with `-race`

Reply:
what tests are failing specifically because of race conditions?

`TestRuleGroupMetrics` is creating all resources in the test function, unless I'm missing something and prometheus has some package level variables under the hood.

Reply:
It feels that tests should not be sharing resources if they are running in parallel. How hard would be to fix the tests? 
Alternatively, maybe you can add an interface and change the implementation of RuleGroupMetrics in runtime by wrapping it with a version that uses mutex.

Reply:
>It feels that tests should not be sharing resources if they are running in parallel. How hard would be to fix the tests?

these tests are not sharing resources. Because starting a the ruler calls `syncRules` and the test itself calls `syncRules` is causing the double invocation. Honestly I don't understand why we need to shy away from this mutex at all. We know it doesn't hurt during run time because syncRules don't really get invoked concurrently.

Reply:
> Honestly I don't understand why we need to shy away from this mutex at all. We know it doesn't hurt during run time because syncRules don't really get invoked concurrently.

because throwing a mutex in the application code just to make a test pass is a weak reason IMHO. It is not for performance, but by the fact that this mutex is essentially not serving its purpose and will mislead people reading this code.

I can think of two options of solving this:
1. Transform `RuleGroupMetrics` into a interface. the implementation can be injected into the ruler by adding a parameter in [this function](https://github.com/cortexproject/cortex/blob/master/pkg/ruler/ruler.go#L297). This will allow you to  pass a thread safe implementation that is a wrapper to the real `RuleGroupMetrics` [here](https://github.com/cortexproject/cortex/blob/master/pkg/ruler/ruler_test.go#L227) by returning the thread safe `RuleGroupMetrics` from `testSetup`, like other entities that are generated only for tests...
2. Add a method to the ruler that will allow us to identify that rules were loaded at least once and we can block and wait in the code. With this approach we will remove any other conditions. 

I prefer option 2 but I fell that this is outside of the scope of this PR, so I think option 1 is ok for this pr.

Reply:
I think that's what tests are for. To detect which parts can breaks with concurrency. The tests proves that this bit of code can break when there is concurrency. That concurrency might not happen in normal execution of the process but the fact remains that it is susceptible to it. I don't like the idea of having an interface and passing a wrapped version in tests because then what is being tested is not the same as what is being used at run time. If someone makes a change to actually call syncRules in parallel the test will not fail. I can try removing the manual call to syncRules and replace it with a wait mechanism

Reply:
updated the test and removed the mtx, PTAL @rapphil , thanks

Reply:
I think tests should test if the behaviour of the code is what is expected and if the assumptions that we make about its behaviour remain constant. If something is not intended to be thread safe, there is no point in testing that. 

Having said that, thanks for fixing the tests :) if one day we need to make `syncRules` thread safe, we should add tests to explicitly test that.